### PR TITLE
Introduce a "Save" button for the admin schedule

### PR DESCRIPTION
**Pull request by Ana06**
The changes were saved after you drop an event, and now it only saves the schedule when you're done with the changes and click the "Save" button.

### CODE REVIEW

File: app/controllers/admin/event_schedules_controller.rb [hunk 1]
```diff
@@ -31,5 +31,11 @@ def destroy
     def event_schedule_params
       params.require(:event_schedule).permit(:schedule_id, :event_id, :room_id, :start_time)
     end
+
+    def parse_errors(event_schedule)
+      title = event_schedule.event.try(:title).present? ? event_schedule.event.title : 'The event'
```
Comment (ChrisBr) [inline app/controllers/admin/event_schedules_controller.rb]: Why / How can that happen?
event_schedule validates presence of event and event validates presence of title, so this case shouldn't happen, should it?

File: app/controllers/admin/event_schedules_controller.rb [hunk 2]
```diff
@@ -31,5 +31,11 @@ def destroy
     def event_schedule_params
       params.require(:event_schedule).permit(:schedule_id, :event_id, :room_id, :start_time)
     end
+
+    def parse_errors(event_schedule)
```
Comment (ChrisBr) [inline app/controllers/admin/event_schedules_controller.rb]: Is this helper function really necessary? I think not ...

**Comment by ChrisBr** [comment]
Also I was able to create two event_schedules for the same event on the same schedule.
Steps to reproduce:

1. Open the schedule in two different browser tabs
2. Schedule on the first tab
3. Schedule on the second tab
4. Save first tab and the second tab
5. Reload the tabs

Result: The event is scheduled on two different times. See attached screenshot.
I see two issues with that: 

1. Something with the JS is broken in this PR
2. We need a validation in the event_schedule.rb model (Please do that in a separate PR!)


schedule

File: app/assets/javascripts/osem-schedule.js [hunk 3]
```diff
@@ -14,66 +17,99 @@ var Schedule = {
...
+  },
+  add: function (previous_parent, new_parent, event) {
+    var params = {
+      previous_parent: previous_parent,
+      new_parent: new_parent,
+      event: event
+    }
+    var event_id = event.attr("event_id");
+    events_to_save[event_id] = params;
+    delete events_to_remove[event_id];
```
Comment (ChrisBr) [inline app/assets/javascripts/osem-schedule.js]: ?
Can we do this somehow different? It's really hard to read/understand!

File: app/assets/javascripts/osem-schedule.js [hunk 4]
```diff
@@ -14,66 +17,99 @@ var Schedule = {
...
-      var error_callback = function(data) {
+      var success_callback_save = function(data) {
+        console.log(data);
+        event.attr("event_schedule_id", data.event_schedule_id);
+        }
+      var error_callback_save = function(data) {
         console.log(data);
-        showError($.parseJSON(data.responseText).errors);
+        errors += $.parseJSON(data.responseText).errors;
+        event.appendTo(events_to_save[key]['previous_parent']);
```
Comment (ChrisBr) [inline app/assets/javascripts/osem-schedule.js]: So in error case it will move the event back to the unscheduled events and shows an error message, right?

**Comment by Ana06** [comment]
@ChrisBr 

> Also I was able to create two event_schedules for the same event on the same schedule.

I know what happened and it is not related to the save feature. When an event is correctly scheduled its `event schedule id `is added to the html as an attribute. So if you have already openned a new tab then the event has no this attribute there and that is why it try to schedule it again. The only way that I think we can avoid that is not using the `event schedule id`, but then we can't use the update and delete method in event_schedule_controller, without id we do not know the route. But I am not sure that we want to do that, at least that will only happens if several organizer are changing the schedule at the same time or if you have several tabs openned. And with the validation we won't have repeated event, only we will get an error when scheduling some of them.

Reloading the page after saving can also be useful to ensure that we have the correct version in case that we had several tabs openned. I can do it in this PR.

File: app/assets/javascripts/osem-schedule.js (hunk 4 above)
Comment (Ana06) [inline app/assets/javascripts/osem-schedule.js]: To unscheduled events or to its previous time and room if it was already scheduled.

File: app/assets/javascripts/osem-schedule.js (hunk 3 above)
Comment (Ana06) [inline app/assets/javascripts/osem-schedule.js]: Why is it hard to understand?

I have two arrays, one for the events that must be saved and another one for those that must be removed. So if you add a event to the `events_to_save` array, it has to be deleted from `events_to_remove`, as you don't want to remove it anymore. I don't understand what is not clear.

File: app/controllers/admin/event_schedules_controller.rb (hunk 1 above)
Comment (Ana06) [inline app/controllers/admin/event_schedules_controller.rb]: It should, but it is an error, if `.save`, `.update` or `.destroy` fails maybe a validation has failed, so I think we can not assume that event and title are present.

**Comment by Ana06** [comment]
I thinking that I can fix it without changing the routes, adding some checks in `event_schedules_controller.rb`

- Unschedule an event twice works.
- Schedule in another time and/or room and event that has already been scheduled twice keep the last one, so I think it is the expected result.
- Schedule an unscheduled event twice fails. With the validation only the first one will be created. I could try to update it when create fails in the controller, so it would keep the second one (I think that is what is expected).
- Schedule in another time and/or room a scheduled event that has been unscheduled in another tab fails (I have realised that currently the javascript doesn't render an error, but it should because rails get a 404 error, so I have to fix that). I could try to create it when update fails in the controller, so it would keep the second one (I think that is what is expected).

File: app/controllers/admin/event_schedules_controller.rb (hunk 2 above)
Comment (Ana06) [inline app/controllers/admin/event_schedules_controller.rb]: Why not? When clicking save I am scheduling several events, so I need to link the error messages to render all together and I use that function to allow that this is done. I add the name of the event (taking into account that validation may have failed), the the errors in brackets and a dot at the end. And I have to do that in `create`, `update` and `destroy`

Look:

image

**Comment by ChrisBr** [comment]
As discussed in IRC: We should introduce batch / bulk :create / :update / :destroy actions. Making a request for each changed event_schedule is ugly and not very high-performance. 

Look for example at this tutorial: http://ericlondon.com/2012/07/18/creating-a-ruby-on-rails-admin-form-to-bulk-update-multiple-models.html

**Comment by Ana06** [comment]
I've introduced the bulk `:create,` `:update` and `:destroy` actions, so every time the save button is clicked only 3 request are done, one per action.

The data saved in `events_to_create`, `events_to_update` and `events_to_remove` hashes use the `event_id` as key. For `events_to_update` I sent a hash which key is the `event_id` and its value is another hash. The key of the second hash is the `event_schedule_id` and the value the params needed to update it. For `events_to_create`  I sent a hash which key is the `event_id` and its value the params needed to create the EventSchedule. Lastly, for `events_to_remove` I sent a hash where the key is the `event_id` and the value the `event_schedule_id` to remove. Basically, if you remove the `event_id` you have what is needed to create, update and delete an `EventSchedule`. So, I don't need to send the `event_id` to the controller but I have decided to send it because for example for `events_to_create` I would like to send an array of objects (with the params), but an array of objects it is sent to the controller like

`{ "0" => {"room_id"=>"25", "start_time"=>"2016-07-14 9:00"}, "1" => {"room_id"=>"3", "start_time"=>"2016-07-14 9:00"}, "2" => {"room_id"=>"7", "start_time"=>"2016-07-14 9:00"}}`

which is a hash where the key is the index and I have to parse it again in the controller, so I directly send the hash with the `event_id`.

**Comment by Ana06** [comment]
I also reload the page after saving the events and I render the error/success message after that.

**Comment by Ana06** [comment]
@ChrisBr @hennevogel @differentreality 
Could you please review this? :birthday:

**Comment by Ana06** [comment]
giphy

**Comment by differentreality** [comment]
@ChrisBr do you wanna finish reviewing this?

**Comment by differentreality** [comment]
Can you please rebase this and take into consideration the changes in https://github.com/openSUSE/osem/pull/1207 ? You have hardcoded some urls in your js that need updating.

**Comment by differentreality** [comment]
Scenario for a little malfunction:
EventA is scheduled at 10:00. Save is fired.
I remove eventA and put in its place eventB (again at 10:00). Save is fired, eventA gets unscheduled, but eventB does not get scheduled.

When save is fired we first fire create (in our case for eventB) and then destroy (in our case for eventA). When we try to schedule eventB, we cannot do that because eventA is still there (since it has not been destroyed yet).

**Comment by Ana06** [comment]
Yes, when I created the PR the replacement feature was not implemented yet, so I didn't take into account.

It'll take me some time to change it so I'll close it and I'll reopen it when it is fixed.

**Comment by differentreality** [comment]
Why do we need to close this? I think only #1168 affects #1082.

**Comment by Ana06** [comment]
But with this the replacement doesn't work, right?

**Comment by differentreality** [comment]
@Ana06 sorry, yes, it's both #1106 and #1168. I left a relevant comment in #1168 about solving it.

**Pull request closed without merging**